### PR TITLE
Fix a CM panic when ipam tries to lock an out of range pre existing cidr

### PR DESCRIPTION
**Pull request by khenidak**
**What type of PR is this?**
 /kind bug

**What this PR does / why we need it**:
fixes: an issue when ipam tries lock a an out of range pre existing node cidr. This usually happen when 
1. User turns on dualstack feature
2. Nodes will get multiple cidrs
3. User turn off dual stack feature
4. Ipam fails to lock pre existing cidrs because they are out of range.

**Which issue(s) this PR fixes**:
Fixes #https://github.com/kubernetes/kubernetes/issues/82765

**Special notes for your reviewer**:
NONE
**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
Similar to changing cluster cidrs. Users are advised to patch node object removing all the cidrs that are no longer apply on cluster when disabling DualStack feature.
```

### CODE REVIEW

**Comment by bowei** [comment]
Can you add a unit test for this case?

**Comment by khenidak** [comment]
@bowei unit tests were added in a separate commit. Thanks.

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 1]
```diff
@@ -284,6 +291,13 @@ func (r *rangeAllocator) ReleaseCIDR(node *v1.Node) error {
 			return fmt.Errorf("failed to parse CIDR %s on Node %v: %v", cidr, node.Name, err)
 		}
 
+		// if node has a pre allocate cidr that does not exist in our cidrs
+		// this will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
+			return fmt.Errorf("node:%s has an allocated cidr: %v at index:%v that does not exist in cluster cidrs configuration. Update node.cidrs by removing this cidr then restart controller manager", node.Name, cidr, idx)
```
Comment (andrewsykim) [inline pkg/controller/nodeipam/ipam/range_allocator.go]: Don't you have to delete the node in this case, not update `--cluster-cidrs` in CM?

Comment (khenidak) [reply pkg/controller/nodeipam/ipam/range_allocator.go]: the reference is `Node.Spec.CIDRs` not `--cluster-cidrs`, will modify

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 2]
```diff
@@ -228,6 +228,13 @@ func (r *rangeAllocator) occupyCIDRs(node *v1.Node) error {
 		if err != nil {
 			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
 		}
+		// if node has a pre allocate cidr that does not exist in our cidrs
+		// this will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
+			return fmt.Errorf("node:%s has an allocated cidr: %v at index:%v that does not exist in cluster cidrs configuration. Update node.cidrs by removing this cidr then restart controller manager", node.Name, cidr, idx)
```
[the remark this answers is not in the document]
Comment (khenidak) [reply pkg/controller/nodeipam/ipam/range_allocator.go]: I have not seen line breaking done elsewhere. Can you point to the guidance?

[the remark this answers is not in the document]
Comment (mogren) [reply pkg/controller/nodeipam/ipam/range_allocator.go]: I agree that it's not common to have to break up lines, and I don't think there are any code guidelines about line length. But I also don't know if I've seen another 237 character wide line in a PR before. Feel free to ignore this comment, but I have seen [error messages that were broken up into multiple lines](https://github.com/kubernetes/kubernetes/blob/c7f9dd0bafe2f3328148a85eed9c18322b9f308e/staging/src/k8s.io/legacy-cloud-providers/vsphere/vclib/diskmanagers/vmdm.go#L84-L86) in other places and that was easier to read.

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 3]
```diff
@@ -284,6 +291,13 @@ func (r *rangeAllocator) ReleaseCIDR(node *v1.Node) error {
 			return fmt.Errorf("failed to parse CIDR %s on Node %v: %v", cidr, node.Name, err)
 		}
 
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
+			return fmt.Errorf("node:%s has an allocated cidr: %v at index:%v that does not exist in cluster cidrs configuration. Update Node.Spec.CIDRs by removing this cidr then restart controller manager", node.Name, cidr, idx)
```
Comment (andrewsykim) [inline pkg/controller/nodeipam/ipam/range_allocator.go]: s/Node.Spec.CIDRs/Node.Spec.PodCIDRs

**Review by andrewsykim — COMMENTED** [review]
One more nit

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 4]
```diff
@@ -228,6 +228,13 @@ func (r *rangeAllocator) occupyCIDRs(node *v1.Node) error {
 		if err != nil {
 			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
 		}
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
+			return fmt.Errorf("node:%s has an allocated cidr: %v at index:%v that does not exist in cluster cidrs configuration. Update Node.Spec.CIDRs by removing this cidr then restart controller manager", node.Name, cidr, idx)
```
Comment (andrewsykim) [inline pkg/controller/nodeipam/ipam/range_allocator.go]: s/Node.Spec.CIDRs/Node.Spec.PodCIDRs

**Comment by andrewsykim** [comment]
@khenidak just realized API validation doesn't allow PodCIDRs to be [updated unless it's empty](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L4505-L4516), does that mean user has to delete node object for this to work?

**Comment by khenidak** [comment]
> updated unless it's empty

As of today yes
I understand what you are trying to say. If i understand correctly. We should say "delete and recreate node object". Which what i don't want to say. We should not tie an error message in a controller to how a type is validated. the type validation will change and folks will absolutely forget to update the error message. The error right now (which btw stops an edge case with a very slim chance of happening, folks don't generally update node cidrs with out doing massive reconfiguration) keeps it as `update the node` how it gets updated, is really up to user how to do it. Btw the original error (in case cluster cidr change and != node.spec.cidrs) is way more cryptic.

**Comment by bowei** [comment]
Unless we change the validation of the node podCIDR, the user is going to have to delete their nodes.
I get not creating a potential textual dependency, but "updating a node" is commonly understood to be `kubectl update node`, which in this case will cause confusion. Since you are giving explicit instructions to the user in the error, might as well give easily understood/actionable one.

**Comment by khenidak** [comment]
How about just removing the part re `please update node.Spec.CIDRs?`

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 5]
```diff
@@ -228,6 +228,13 @@ func (r *rangeAllocator) occupyCIDRs(node *v1.Node) error {
 		if err != nil {
 			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
 		}
+		// if node has a pre allocate cidr that does not exist in our cidrs
+		// this will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
```
[the remark this answers is not in the document]
Comment (mogren) [reply pkg/controller/nodeipam/ipam/range_allocator.go]: I wouldn't say that this was resolved... The comment you have is hard to make sense of. For example, what does "we have now way" on the last line even mean?

**Comment by bowei** [comment]
thank you

/approve
/lgtm

**Pull request merged**